src/main/kotlin/com/exactpro/th2/rptdataprovider/handlers/messages/MessageExtractor.kt

```diff
@@ -138,139 +100,144 @@ class MessageExtractor<B, G, RM, PM>(
+                )
+                val cradleMessageIterable = context.cradleService.getGroupedMessages(
+                    this,
+                    GroupedMessageFilterBuilder().apply {
+                        bookId(BOOK_ID)
+                        groupName(sessionGroup)
+                        order(order)
+
+                        if (order == REVERSE) { // default: DIRECT
+                            request.startTimestamp?.let { timestampTo().isLessThanOrEqualTo(it) }
+                            request.endTimestamp?.let { timestampFrom().isGreaterThan(it) }
+                        } else {
+                            request.startTimestamp?.let { timestampFrom().isGreaterThanOrEqualTo(it) }
+                            request.endTimestamp?.let { timestampTo().isLessThan(it) }
+                        }
+                    }.build()
                 )
 
-                logger.debug { "cradle iterator has been built for stream $streamName" }
-
-                var isLastMessageTrimmed = false
+                LOGGER.debug { "cradle iterator has been built for session group: $sessionGroup, alias: $commonStreamName" }
 
-                for (batch in cradleMessageIterable) {
-                    if (externalScope.isActive && !isLastMessageTrimmed) {
+                var lastNotFound = true
 
+                for (batch: StoredGroupedMessageBatch in cradleMessageIterable) {
+                    if (externalScope.isActive && lastNotFound) {
                         val timeStart = System.currentTimeMillis()
 
-                        pipelineStatus.fetchedStart(streamName.toString())
-                        logger.trace { "batch ${batch.id.sequence} of stream $streamName with ${batch.messageCount} messages (${batch.batchSize} bytes) has been extracted" }
+                        pipelineStatus.fetchedStart(commonStreamName.toString())
+                        LOGGER.trace { "batch ${batch.firstTimestamp} of group $sessionGroup for stream $commonStreamName with ${batch.messageCount} messages (${batch.batchSize} bytes) has been extracted" }
+
+                        val resumeSequence = resumeFromId?.sequence
+                        val start = request.startTimestamp
+                        val end = request.endTimestamp
 
-                        val trimmedMessages = trimMessagesListHead(getMessagesFromBatch(batch), resumeFromId)
-                            .dropLastWhile { message ->
-                                trimMessagesListTail(message).also {
-                                    isLastMessageTrimmed = isLastMessageTrimmed || it
+                        val trimmedMessages = batch.orderedMessages.asSequence()
+                            .filter { it.sessionAlias == commonStreamName.name }
+                            .dropWhile(
+                                when {
+                                    resumeSequence != null -> { { sequenceComparator(it.sequence, resumeSequence) } }

```

Comment:
issue: we can't use sequence without direction of message id because the stream related to session group can contains messages for both direction where sequences are independent.
Please add the next test:
 user would like to continue from id book:alias:<time>:FIRST:2 and process in direct order
 stream contains the next messages:
 book:alias:<time>:SECOND:1
 book:alias:<time>:SECOND:2
 book:alias:<time>:FIRST:1
 book:alias:<time>:SECOND:3
 book:alias:<time>:FIRST:2 // point for continue 
 book:alias:<time>:FIRST:3